cvat-cli/src/cvat_cli/_internal/agent.py

```diff
@@ -0,0 +1,337 @@
+            max_workers=1,
+            mp_context=self._mp_context,
+            initializer=self._initializer,
+            initargs=self._initargs,
+        )
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        self._executor.shutdown()
+
+    def submit(self, func, /, *args, **kwargs):
+        return self._executor.submit(func, *args, **kwargs)
+
+    def result(self, future: concurrent.futures.Future):
+        try:
+            return future.result()
+        except concurrent.futures.BrokenExecutor:
+            self._executor.shutdown()
+            self._executor = self._new_executor()
+            raise
+
+
+_current_function: cvataa.DetectionFunction
+
+
+def _worker_init(function_loader: FunctionLoader):
+    global _current_function
+    _current_function = function_loader.load()
+
+
+def _worker_job_get_function_spec():
+    return _current_function.spec
+
+
+def _worker_job_detect(context, image):
+    return _current_function.detect(context, image)
+
+
+class _Agent:
+    def __init__(self, client: Client, executor: _RecoverableExecutor, function_id: int):
+        self._client = client
+        self._executor = executor
+        self._function_id = function_id
+        self._function_spec = self._executor.result(
+            self._executor.submit(_worker_job_get_function_spec)
+        )
+
+        _, response = self._client.api_client.call_api(
+            "/api/functions/{function_id}",
+            "GET",
+            path_params={"function_id": self._function_id},
+        )
+
+        remote_function = json.loads(response.data)
+
+        self._validate_function_compatibility(remote_function)
+
+        self._agent_id = secrets.token_hex(16)
+        self._client.logger.info("Agent starting with ID %r", self._agent_id)
+
+        self._cached_task_id = None
+
+    def _validate_function_compatibility(self, remote_function: dict) -> None:
+        function_id = remote_function["id"]
+
+        if remote_function["provider"] != "native":
+            raise CriticalError(
+                f"Function #{function_id} has provider {remote_function['provider']!r}. "
+                "Agents can only be run for functions with provider 'native'."
+            )
+
+        if isinstance(self._function_spec, cvataa.DetectionFunctionSpec):
+            self._validate_detection_function_compatibility(remote_function)
+            self._calculate_result_for_ar = self._calculate_result_for_detection_ar
+        else:
+            raise CriticalError(
+                f"Unsupported function spec type: {type(self._function_spec).__name__}"
+            )
+
+    def _validate_detection_function_compatibility(self, remote_function: dict) -> None:
+        incompatible_msg = (
+            f"Function #{remote_function['id']} is incompatible with function object: "
+        )
+
+        if remote_function["kind"] != "detector":
```

Comment:
Consider adding a named constant or enum for `detector`.

Reply:
Added.